### DISCUSSION TITLE
State of skipped scenarios in JSON report

### DISCUSSION THREAD

**Discussion by csvtuda** [badeball/cypress-cucumber-preprocessor#1203] [Q&A] [answered]
Hi,

first of all, thanks for the great plugin. I maintain a [plugin](https://www.npmjs.com/package/cypress-xray-plugin) for uploading test run results to a test management tool and have recently been asked why tests marked as skipped (`@skip`) do not show up in the uploaded results (see https://github.com/Qytera-Gmbh/cypress-xray-plugin/issues/341). My plugin relies heavily on the JSON report that your your plugin outputs and uploads more or less as is. 

A MWE revealed that skipped tests do not show up at all in the produced report:

```gherkin
# cucumber.feature
Feature: Testing skips

    @skip
    Scenario: Scenario 1
        Given a step

    Scenario: Scenario 2
        Given a step
```

```jsonc
// cucumber-report.json
[
  {
    "description": "",
    "elements": [
      {
        "description": "",
        "id": "testing-skips;scenario-2",
        "keyword": "Scenario",
        "line": 4,
        "name": "Scenario 2",
        "steps": [
          {
            "arguments": [],
            "keyword": "Given ",
            "line": 5,
            "name": "a step",
            "match": {
              "location": "not available:0"
            },
            "result": {
              "status": "passed",
              "duration": 26000000
            }
          }
        ],
        "tags": [],
        "type": "scenario"
      }
    ],
    "id": "testing-skips",
    "line": 1,
    "keyword": "Feature",
    "name": "Testing skips",
    "tags": [],
    "uri": "cucumber.feature"
  }
]
```

---

I then searched this repository for any issues discussing this topic. If I understand correctly, the timeline was as follows:

1. skipped scenarios were not initially included
2. [#329](https://github.com/badeball/cypress-cucumber-preprocessor/pull/329) was merged, which then added skipped tests back into the reports and introduced [undesired side effects](https://github.com/badeball/cypress-cucumber-preprocessor/pull/329#issuecomment-654115643)
4. mitigation attempt: #474 
5. mitigation attempt: #485

There's talks about reverting #329 in the mitigation attempts, but I haven't found any more information on that. I assume the changes have been reverted at some point?

---

Question: What's the final decision on how and if skipped tests should be included in the JSON report? I'm not asking for a feature, just for clarification so I can handle my own issue accordingly.

**Comment by badeball** [comment, marked as answer]
> What's the final decision on how and if skipped tests should be included in the JSON report?

Skipped tests should ideally be included in the JSON report and marked as such. However, this is technically unfeasible without also introducing the mentioned undesired side effects.

Lack of status of skipped tests also affects HTML reports, which is reported [here](https://github.com/badeball/cypress-cucumber-preprocessor/issues/1041).

This would ideally be solved with more fine-grained plugin events, which has been [previously dicussed](https://github.com/cypress-io/cypress/issues/2283), but never made it upstream. There is a [proposal](https://github.com/cypress-io/cypress/issues/27175), but as explained in https://github.com/badeball/cypress-cucumber-preprocessor/issues/1041#issuecomment-1599430189, I don't think anything will ever happen with this, simply because Cypress hasn't made any meaningful, positive changes towards plugin developers in a long time.

**Reply by csvtuda** [reply]
Thanks for the quick reply and the further related links. It's unfortunate, but I can obviously instead leverage Cypress's own run results, which include the test in its (extra confusing...) `pending` state:

```json
"tests": [
  {
    "attempts": [
      {
        "state": "pending"
      }
    ],
    "displayError": null,
    "duration": 39,
    "state": "pending",
    "title": [
      "Testing skips",
      "Scenario 1"
    ]
  }
],
```

So I guess I'll be joining the waitlist of your proposal. Fingers crossed for some updates in that regard.

**Reply by badeball** [reply]
I couldn't quite shake this away, so I decided to take another stab at it and made it work despite the lack of fine-grained plugin events. Skipped scenarios (however skipped) are now included in the reports (hopefully). I've released this as v20.1.0.

**Reply by csvtuda** [reply]
After a little bit of experimentation, the new version seems to reliably include the skipped scenarios as well, thanks a lot!
